_pytest/python_api.py

```diff
@@ -6,10 +6,271 @@
+        one element is ambiguous. Use a.any() or a.all()" when compared with a 
+        custom class, so `b.__eq__` never gets called.
+
+        The trick is that the priority rules change if `a` and `b` are related 
+        by inheritance.  Specifically, `b.__eq__` gets priority if `b` is a 
+        subclass of `a`.  So we can guarantee that `ApproxNumpy.__eq__` gets 
+        called by inheriting from `numpy.ndarray`.
+        """
+
+        def __new__(cls, expected, rel=None, abs=None):
+            """
+            Numpy uses __new__ (rather than __init__) to initialize objects. 
+            
+            The `expected` argument must be a numpy array.  This should be 
+            ensured by the approx() delegator function.
+            """
+            assert isinstance(expected, np.ndarray)
+            obj = super(ApproxNumpy, cls).__new__(cls, expected.shape)
+            obj.__init__(expected, rel, abs)
+            return obj
+
+        def __repr__(self):
+            # It might be nice to rewrite this function to account for the 
+            # shape of the array...
+            return '[' + ApproxBase.__repr__(self) + ']'
+
+        def __eq__(self, actual):
+            try:
+                actual = np.array(actual)
+            except:
+                raise ValueError("cannot cast '{0}' to numpy.ndarray".format(actual))
+
+            if actual.shape != self.expected.shape:
+                return False
+
+            return ApproxBase.__eq__(self, actual)
+
+        def _yield_expected(self):
+            for x in self.expected:
+                yield x
+
+        def _yield_comparisons(self, actual):
+            # We can be sure that `actual` is a numpy array, because it's 
+            # casted in `__eq__` before being passed to `ApproxBase.__eq__`, 
+            # which is the only method that calls this one.
+            for i in np.ndindex(self.expected.shape):
+                yield actual[i], self.expected[i]
+
+
+except ImportError:
+    np = None
+
+class ApproxMapping(ApproxBase):
+    """
+    Perform approximate comparisons for mappings where the values are numbers 
+    (the keys can be anything).
+    """
+
+    def __repr__(self):
+        item = lambda k, v: "'{0}': {1}".format(k, self._approx_scalar(v))
+        return '{' + ', '.join(item(k,v) for k,v in self.expected.items()) + '}'
+
+    def __eq__(self, actual):
+        if actual.keys() != self.expected.keys():
+            return False
+
+        return ApproxBase.__eq__(self, actual)
+        
+    def _yield_comparisons(self, actual):
+        for k in self.expected.keys():
+            yield actual[k], self.expected[k]
+
+
+class ApproxSequence(ApproxBase):
+    """
+    Perform approximate comparisons for sequences of numbers.
+    """
+
+    def __repr__(self):
+        open, close = '()' if isinstance(self.expected, tuple) else '[]'
+        return open + ApproxBase.__repr__(self) + close
```

Comment:
Watch out because tuples of one element should end with a `,`, like `(1.0,)`.
Suggestion:

```
return type(self.expected)(self._approx_scalar(x) for x in self._yield_expected())
```